NBitcoin/Scripting/OutputDescriptorParser.cs

```diff
@@ -0,0 +1,303 @@
+			from item in pInner
+			from _r in Parse.Char(')')
+			select constructor(item);
+
+		private static P PPKHelper(string name, Func<PubKeyProvider, OutputDescriptor> constructor, ISigningRepository? repo, bool onlyCompressed) =>
+			PExprHelper<PubKeyProvider>(Parse.String(name).Text(), PPubKeyProvider(repo, onlyCompressed), constructor);
+
+		internal static P PPK(ISigningRepository? repo, bool onlyCompressed = false) =>
+			PPKHelper("pk", OutputDescriptor.NewPK, repo, onlyCompressed);
+
+		internal static P PPKH(ISigningRepository? repo, bool onlyCompressed = false) =>
+			PPKHelper("pkh", OutputDescriptor.NewPKH, repo, onlyCompressed);
+
+		internal static P PWPKH(ISigningRepository? repo) =>
+			PPKHelper("wpkh", OutputDescriptor.NewWPKH, repo, true);
+
+		internal static P PCombo(ISigningRepository? repo, bool onlyCompressed = false) =>
+			PPKHelper("combo", OutputDescriptor.NewCombo, repo, onlyCompressed);
+
+		internal static P PMulti(ISigningRepository? repo, bool onlyCompressed, uint? maxN = null) =>
+			from name in Parse.String("sortedmulti").XOr(Parse.String("multi")).Text()
+			let isSorted = name.StartsWith("sorted")
+			from _l in Parse.Char('(')
+			from m in Parse.Digit.XMany().Text().Then(d => Parse.TryConvert(d.Trim(), UInt32.Parse))
+			where m != 0
+			from _c in Parse.Char(',')
+			from pkProviders in PPubKeyProvider(repo, onlyCompressed).DelimitedBy(Parse.Char(','))
+			where m <= pkProviders.Count()
+			from _r in Parse.Char(')')
+			where !maxN.HasValue || pkProviders.Count() <= maxN
+			select OutputDescriptor.NewMulti(m, pkProviders, isSorted);
+
+		internal static P PWSHInner(ISigningRepository? repo, bool onlyCompressed = false, uint? maxMultisigKeyN = null) =>
+			PPK(repo, onlyCompressed)
+				.Or(PPKH(repo, onlyCompressed))
+				.Or(PMulti(repo, onlyCompressed, maxMultisigKeyN));
+		internal static P PInner(ISigningRepository? repo, bool onlyCompressed = false, uint? maxMultisigKeyN = null) =>
+			PWSHInner(repo, onlyCompressed, maxMultisigKeyN)
+				.Or(PWPKH(repo));
+
+		internal static P PWSH(ISigningRepository? repo) =>
+			PExprHelper(Parse.String("wsh").Text(), PWSHInner(repo, true), OutputDescriptor.NewWSH);
+		internal static P PSH(ISigningRepository? repo) =>
+			PExprHelper((Parse.String("sh").Text()).Except(Parse.String("wsh")), PInner(repo, false, 15).Or(PWSH(repo)), OutputDescriptor.NewSH);
+
+		internal static P POutputDescriptor(ISigningRepository? repo) =>
+			PAddr
+				.Or(PRaw(repo))
+				.Or(PCombo(repo))
+				.Or(PInner(repo))
+				.Or(PWSH(repo))
+				.Or(PSH(repo)).End();
+
+		internal static bool TryParseOD(string str, out OutputDescriptor? result, bool requireCheckSum = false, ISigningRepository? repo = null)
+			=> TryParseOD(str, out _, out result,requireCheckSum, repo);
+		private static bool TryParseOD(string str, out string? whyFailure, out OutputDescriptor? result, bool requireCheckSum = false, ISigningRepository? repo = null)
+		{
+			str = str.Replace(" ", "");
```

Comment:
NRE here if the caller pass `null`. Given it is not possible to prevent callers to pass null I think it is a good idea to check for null on the public interface.